semistruct.go

```diff
@@ -0,0 +1,283 @@
+package semistruct
+
+import "strconv"
+import p "github.com/andyleap/parser"
+import "errors"
+
+// Log is the data structure representing the result of a successful
+// semistructured log line parse.
+type Log struct {
+	Priority int64
+	Tags     []string
+	Attrs    map[string]string
+}
+
+type kvPair struct {
+	key string
+	val string
+}
+
+// NewLogParser instantiates a parser composed of combinators for
+// parsing a semistructured log line into the SemistructLog struct
+// data type.
+func NewLogParser() *p.Grammar {
+	skipSpace := SkipSpace()
+
+	o := p.And(
+		// Parse opening sentinel "!<"
+		OpenSentinel(),
+		skipSpace,
+
+		// Parse our log-level priority
+		p.Tag("priority", PriorityInt()),
+		skipSpace,
+
+		// Parse our tag list "[tag1:tag2:tag3]"
+		p.Tag("tags", Tags()),
+		skipSpace,
+
+		// Parse our attribute set "{ key=val key2=val2 }"
+		p.Tag("attrs", Attrs()),
+		skipSpace,
+
+		// Parse the ending sentinel ">!"
+		EndSentinel(),
+	)
+
+	o.Node(func(m p.Match) (p.Match, error) {
+
+		// If we're parsing, we absolutely want a priority - if that
+		// returns an error then we should return an error; the other
+		// types can have sane defaults.
+		pr, ok := p.GetTag(m, "priority").(int64)
+		if !ok {
+			return nil, errors.New("failed getting priority")
```

Comment:
Yeah I checked https://golang.org/doc/effective_go.html search for `ErrInternal` and `net/http`. Most of the go standard library appears to use this format.